### DISCUSSION TITLE
Building shared object on Windows with MSYS2/UCRT64

### DISCUSSION THREAD

**Discussion by mtsokol** [mesonbuild/meson#13613] [Q&A] [answered]
Hi!

We recently ported OpenBLAS project to Meson build system (it's following Makefiles for a reference). We support Linux and MacOS based systems, and `x86_64` and `arm64` architectures.

Recently I worked on adding Windows support, using MSYS2/UCRT64 environment: https://github.com/HaoZeke/OpenBLAS/pull/26 (only added new CI job TBH).

To build static and dynamic libraries we use `both_libraries(...)`:

```meson
openblas = both_libraries('openblas',
  link_whole: [
    _interface,
    _l2_driver,
    _l3_driver,
    _others,
    _kern,
  ],
  dependencies: [dependency('threads')],
  override_options: ['b_lundef=false', 'b_asneeded=false'],
  install: true,
)
```

 It turns out that for Windows building dynamic library (meson command below) fails with multitude of `undefined reference to`:

```
 FAILED: libopenblas.dll 
"cc"  -o libopenblas.dll  "-LC:/msys64/ucrt64/lib/gcc/x86_64-w64-mingw32/14.2.0" 
"-LC:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0" 
"-LC:/msys64/ucrt64/lib/gcc" "-LC:/msys64/ucrt64/bin/../lib/gcc" "-LC:/msys64/ucrt64/x86_64-w64-mingw32/lib" 
"-LC:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/lib/../lib" 
"-LC:/msys64/ucrt64/lib" "-LC:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../lib" 
"-LC:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/lib" 
"-LC:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../.." "-Wl,--allow-shlib-undefined" "-Wl,-O1" 
"-shared" "-Wl,--start-group" "-Wl,--out-implib=libopenblas.dll.a" "-Wl,--whole-archive" 
"interface/lib_interface.a" "driver/level2/libl2_driver.a" "driver/level3/libl3_driver.a" 
"driver/others/lib_interface.a" "kernel/lib_kern.a" "-Wl,--no-whole-archive" "-pthread" "-lkernel32" 
"-luser32" "-lgdi32" "-lwinspool" "-lshell32" "-lole32" "-loleaut32" "-luuid" "-lcomdlg32" "-ladvapi32" 
"-lgfortran" "-lm" "-Wl,--end-group"
```

It's the first time I'm using MSYS2, I wonder why MSYS2 gcc build would be different from linux gcc/clang build that manages to build `so`/`dylib` (or is it meson+windows specific behavior that I should handle differently?).

I'd be thankful for any tips, comments how to debug it!

**Comment by mattip** [comment]
Where is the actual error message?

**Comment by mattip** [comment]
Ahh, this is on "discussions". I will continue on the PR itself.

**Comment by mtsokol** [comment]
Sure! Here's the PR with failing windows job: https://github.com/HaoZeke/OpenBLAS/pull/26

**Comment by HaoZeke** [comment, marked as answer]
@mtsokol this can be closed (the original issue had to do with incorrectly resolved symbols at our end).

**Comment by mtsokol** [comment]
@HaoZeke Sure! Thanks!
